### PR TITLE
Transaction pool: Ensure that we prune transactions properly

### PR DESCRIPTION
**Pull request by bkchr**
There was a bug in the transaction pool that we didn't pruned
transactions properly because we called `prune_known`, instead of `prune`.

This bug was introduced by:
https://github.com/paritytech/substrate/pull/4629

This is required to have stale extrinsics being removed properly, so
that they don't fill up the tx pool.

### CODE REVIEW

**Review by hendis85 — APPROVED** [review]

File: client/transaction-pool/src/lib.rs [hunk 1]
```diff
@@ -681,6 +696,15 @@ impl<PoolApi, Block> MaintainedTransactionPool for BasicPool<PoolApi, Block>
 						move || Box::new(extra_pool.validated_pool().ready()),
 					);
 
+					// if let Err(e) = pool.validated_pool().clear_stale(&id) {
+					// 	log::debug!(
+					// 		target: "txpool",
+					// 		"Error while clearing stale transaction at {:?}: {:?}",
```
Comment (tomusdrw) [inline client/transaction-pool/src/lib.rs:702]: Commented code?

**Review by tomusdrw — COMMENTED** [review]
Looks reasonable other than the comments.

File: client/transaction-pool/graph/src/listener.rs [hunk 2]
```diff
@@ -99,12 +99,8 @@ impl<H: hash::Hash + traits::Member + Serialize, C: ChainApi> Listener<H, C> {
 	}
 
 	/// Transaction was removed as invalid.
-	pub fn invalid(&mut self, tx: &H, warn: bool) {
-		if warn {
-			warn!(target: "txpool", "[{:?}] Extrinsic invalid", tx);
-		} else {
-			debug!(target: "txpool", "[{:?}] Extrinsic invalid", tx);
-		}
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/transaction-pool/graph/src/listener.rs:107]: Nobody ever looked at these warnings, especially as they were also triggered by re-validating inherents for example.

If we want to improve this, we should print proper messages where this function is called, but not down here where we have no idea what actually happens.

[the remark this answers is not in the document]
Comment (tomaka) [reply client/transaction-pool/graph/src/listener.rs:107]: > Nobody ever looked at these warnings

Indeed. The problem is that these warnings are not some rare event but get printed extremely often. At the beginning people were concerned, but over time they have learned to completely ignore them.

**Review by gavofyork — APPROVED** [review]

**Review by andresilva — APPROVED** [review]

File: client/transaction-pool/src/lib.rs [hunk 3]
```diff
@@ -555,19 +557,32 @@ async fn prune_known_txs_for_block<Block: BlockT, Api: ChainApi<Block = Block>>(
...
+			return hashes
+		},
+		Err(e) => {
+			log::debug!(target: "txpool", "Error retrieving header for {:?}: {:?}", block_id, e);
+			return hashes
+		}
+	};
+
+	if let Err(e) = pool.prune(&block_id, &BlockId::hash(*header.parent_hash()), &extrinsics).await {
 		log::error!("Cannot prune known in the pool {:?}!", e);
```
Comment (gilescope) [inline client/transaction-pool/src/lib.rs:586]: should the error line be updated to not say `prune known`?

**Pull request merged**